directembedding/src/main/scala/ch/epfl/directembedding/transformers/EmbeddingTransformation.scala

```diff
@@ -0,0 +1,258 @@
+      }
+
+      reification match {
+        case Nil =>
+          c.abort(tree.pos, s"$sym on ${selfType(tree).typeSymbol} is not supported in $dslName")
+        case r :: Nil =>
+          ReifyTransform(r)
+        case _ =>
+          c.abort(tree.pos, s"Ambiguous reification with alteratives:\n${reification.mkString("  ", "\n", "")}")
+      }
+    }
+
+    private def isNotModule(tree: Tree): Boolean = {
+      val expr = tree match {
+        // Typed blocks have no symbol
+        case Typed(lhs, typ) => lhs
+        case Block(_, rhs)   => rhs
+        case _               => tree
+      }
+      expr.symbol != null && !expr.symbol.isModule
+    }
+
+    /**
+     * Convert curried function to uncurried function.
+     *
+     * Example f(1)(2) => f(1, 2)
+     *
+     * @param t Tree to be uncurried
+     * @param args List[Tree] arguments to prepend to uncurried argument list
+     * @return Uncurried tree
+     */
+    private def uncurry(t: Tree, args: List[Tree]): Tree = t match {
+      case field @ Apply(x, y) =>
+        uncurry(x, y ::: args)
+      case _ =>
+        Apply(t, args)
+    }
+
+    /**
+     * Get correct symbol of invoked member.
+     *
+     * Handles the case when the member is a val, which scalac translates
+     * into a private member.
+     * @param field
+     * @return
+     */
+    private def invokedSymbol(field: Select): Symbol = {
+      val reifies = field.symbol.annotations.find(_.tree.tpe <:< c.typeOf[DirectEmbeddingAnnotation])
+      reifies.map(_ => field.symbol).getOrElse { // find annotations on desugared vals
+        field.symbol.owner.info.members.find(equalSymbols(field.symbol)).getOrElse {
+          c.abort(field.pos, s"Unable to find symbol for val $field.")
+        }
+      }
+    }
+
+    /**
+     * Get body of invoked instance.
+     *
+     * Example: {{{
+     *   self(q"Predef.implicitly[T]") == None
+     *   self(q"new Foo()") == Some(q"Foo")
+     * }}}
+     */
+    private def self(lhs: Tree): Option[Tree] = lhs match {
+      case Select(x, _) if isNotModule(x) => Some(x)
+      case TypeApply(Select(x, _), _) if isNotModule(x) => Some(x)
+      case _ => None
+    }
+
+    /**
+     * Get types that are embedded inside the left-hand side of the application.
+     *
+     * Example: {{{
+     *   hiddenTypeArgs(q"new TArgClassExample[Int]") == List(tq"Int")
+     * }}}
+     */
+    private def hiddenTypeArgs(lhs: Tree): List[Tree] = lhs match {
+      case Select(x, _)        => x.tpe.typeArgs.map(TypeTree(_))
+      case TypeApply(_, targs) => targs
+      case _                   => Nil
+    }
+
+    private def symbol(lhs: Tree): Symbol = lhs match {
+      case Select(New(newBody), _)     => newBody.symbol
+      case field: Select               => invokedSymbol(field)
+      case TypeApply(field: Select, _) => invokedSymbol(field)
+      case _                           => lhs.symbol
+    }
+
+    private def selfType(tree: Tree): Type = {
```

Comment:
This method may contain a bug, if you can find a counterexample I am happy to add an issue.